### ISSUE TITLE
Full configurability of the Microsoft provider (and potentially others) via the admin panel

### ISSUE THREAD

**Issue by Antvirf** [pennersr/django-allauth#3424]
Filing this issue to discuss a potential implementation / suggestion (or perhaps hear an 'obvious' way to solve this easily...) for full configurability of the Microsoft provider via admin panel settings. Specifically I'm here referring to a configuration for the Microsoft provider that is **single-tenant** in MS/Azure terms, i.e. only users from the Azure AD instance that created the app configuration can log in.

Happy to try and put together a PR to address but would prefer to get some input first on the best way to do this.

## Scenario

- Allauth 0.56.0, Django 4.2.4, Python 3.9.7
- Single code base with a single `settings.py` file/module
- Multi-tenant application, in which each tenant has their own allauth application installed
- Each tenant may have a different auth provider configured via allauth from a selected shortlist
- The authentication configuration for each tenant has to be set via the admin panel, otherwise the `settings.py` of the project has to configure the provider for each tenant (i.e. requiring code changes to add or reconfigure the auth method)

## Challenge

The Microsoft provider expects a `TENANT` argument provided via `settings.py` to correctly configure the authentication endpoint URL for use with a "single-tenant" AD application, see this [Stack Overflow answer for reference](https://stackoverflow.com/a/73647651).

The [relevant part](https://github.com/pennersr/django-allauth/blob/ba92ef383865ea3f529fa9b972ee615756b573fa/allauth/socialaccount/providers/microsoft/views.py#L40) of the Microsoft provider code can be found in the views.py:

```python
...
class MicrosoftGraphOAuth2Adapter(OAuth2Adapter):
    provider_id = MicrosoftGraphProvider.id

    settings = app_settings.PROVIDERS.get(provider_id, {})
    # Lower case "tenant" for backwards compatibility
    tenant = settings.get("TENANT", settings.get("tenant", "common"))

    provider_base_url = "https://login.microsoftonline.com/{0}".format(tenant)
...
```

For a "single-tenant" AD application, we have to provide the `TENANT` setting. Following the definitions/code and looking at [`socialaccount/app_settings.py`](https://github.com/pennersr/django-allauth/blob/ba92ef383865ea3f529fa9b972ee615756b573fa/allauth/socialaccount/app_settings.py) and [`utils.py`](https://github.com/pennersr/django-allauth/blob/ba92ef383865ea3f529fa9b972ee615756b573fa/allauth/utils.py), it seems that only the settings configured via code are applied. Configuring `TENANT` from the admin panel does not seem to work - the value is ignored, and the provider falls back to `/common/` as the default, resulting in auth failure.

## Perfect world scenario

Optimally, the `TENANT` argument could be provided via the settings field:

![image](https://github.com/pennersr/django-allauth/assets/26534322/304770ea-794f-4032-a595-115ec38fdaad)

**Comment by pennersr** [comment]
Indeed, it would make for a good change to source the tenant from the app settings. The reason that this hasn't been done from the start is merely that the app-level settings are only recently introduced.

Likely, you can replace the values that require the tenant, like `access_token_url`, `authorize_url` and `profile_url` with something like (warning -- untested code sketch):

```
@property 
def access_token_url(self):
    app = get_adapter().get_app(context.request, self.provider_id)
    settings = app_settings.PROVIDERS.get(provider_id, {})
    tenant = settings.get("TENANT", settings.get("tenant", "common"))
    tenant = app.settings.get("tenant", tenant)  # Lower case is intentional here.
    return "https://login.microsoftonline.com/{0}".format(tenant)
```

**Comment by Antvirf** [comment]
Thanks @pennersr, makes sense. I'll do some playing around with the snippet you provided and see if I can find an easy way around this for now.

For a broader change, what's your view in terms of precedence for potentially overlapping settings keys? My feeling is that whatever an admin sets via the admin panel is probably what would be more 'recent' / relevant / up to date and should take precedence over code.

**Comment by derek-adair** [comment]
This would be a positive change for sure.  

Agreed that manually set settings should override any configured at runtime.

**Comment by pennersr** [comment]
Closed via cb6ef5d0

**Comment by pennersr** [comment]
https://codeberg.org/allauth/django-allauth/issues/3424
